entities/storobj/storage_object.go

```diff
@@ -1131,6 +1169,55 @@ func unmarshalTargetVectors(rw *byteops.ReadWriter) (map[string][]float32, error
 	return nil, nil
 }
 
+// TODO doc for onlyUnmarshalNames
+func unmarshalMultiVectors(rw *byteops.ReadWriter, onlyUnmarshalNames map[string]interface{}) (map[string][][]float32, error) {
+	// TODO is this comment still right? DRY comment/code?
+	// This check prevents from panic when somebody is upgrading from version that
+	// didn't have multi vector support. This check is needed bc with named vectors
+	// feature storage object can have vectors data appended at the end of the file
+	if rw.Position < uint64(len(rw.Buffer)) {
+		multiVectorsOffsets := rw.ReadBytesFromBufferWithUint32LengthIndicator()
+		multiVectorsSegmentLength := rw.ReadUint32()
+		pos := rw.Position
+
+		if len(multiVectorsOffsets) > 0 {
+			var mvOffsets map[string]uint32
+			if err := msgpack.Unmarshal(multiVectorsOffsets, &mvOffsets); err != nil {
+				return nil, fmt.Errorf("Could not unmarshal multi vectors offset: %w", err)
+			}
+
+			// NOTE if you sort mvOffsets by offset, you may be able to speed this up via
+			// sequential reads, haven't tried this yet
+			multiVectors := map[string][][]float32{}
+			for name, offset := range mvOffsets {
+				// if onlyUnmarshalNames is set and non-empty, only unmarshal the vectors
+				// for the names in the map
+				if onlyUnmarshalNames != nil && len(onlyUnmarshalNames) > 0 {
+					if _, ok := onlyUnmarshalNames[name]; !ok {
+						continue
+					}
+				}
```

Comment:
This block makes the behavior of `unmarshalMultiVectors` different from `unmarshalTargetVectors`:
- `unmarshalMultiVectors`: If vector names are provided, only unmarshals those vectors and not the others.
- `unmarshalTargetVectors`: Always unmarshals all vectors, even if called from `FromBinaryOptional`

Is this too confusing? Should I modify them to match or make a new func for multi vectors?

Reply:
Marcin says leave the target vector behavior the same as before, no strong opinion on the multi vector function